x-pack/plugin/enrich/src/main/java/org/elasticsearch/xpack/enrich/EnrichPolicyRunner.java

```diff
@@ -0,0 +1,298 @@
+                if (properties.containsKey(enrichField) == false) {
+                    listener.onFailure(new ElasticsearchException(
+                        "Enrich policy execution for [{}] failed. Could not locate enrich value field [{}] on mapping for index [{}]",
+                        policyName, enrichField, sourceIndex));
+                }
+            }
+        }
+    }
+
+    private String getEnrichIndexBase(final String policyName) {
+        return ".enrich-" + policyName;
+    }
+
+    private Map<String, Object> resolveEnrichMapping(final EnrichPolicy policy) {
+        // TODO: Modify enrichKeyFieldMapping based on policy type.
+        // Create dynamic mapping templates - one for the enrich key and one for the enrich values
+        Map<String, Object> enrichKeyFieldMapping = MapBuilder.<String, Object>newMapBuilder()
+            .put("type", "{dynamic_type}")
+            .put("indexed", true)
+            .put("doc_values", false)
+            .map();
+
+        Map<String, Object> enrichKeyTemplateDefinition = MapBuilder.<String, Object>newMapBuilder()
+            .put("match", policy.getEnrichKey())
+            .put("mapping", enrichKeyFieldMapping)
+            .map();
+
+        Map<String, Object> enrichKeyTemplate = MapBuilder.<String, Object>newMapBuilder()
+            .put("enrich_key_template", enrichKeyTemplateDefinition)
+            .map();
+
+        Map<String, Object> enrichTextValueFieldMapping = MapBuilder.<String, Object>newMapBuilder()
+            .put("type", "keyword")
+            .put("ignore_above", 256)
+            .put("indexed", false)
+            .put("doc_values", true)
+            .map();
+
+        Map<String, Object> enrichTextValueTemplateDefinition = MapBuilder.<String, Object>newMapBuilder()
+            .put("match_mapping_type", "string")
+            .put("match", "*")
+            .put("unmatch", policy.getEnrichKey())
+            .put("mapping", enrichTextValueFieldMapping)
+            .map();
+
+        Map<String, Object> enrichTextValueTemplate = MapBuilder.<String, Object>newMapBuilder()
+            .put("enrich_text_value_template", enrichTextValueTemplateDefinition)
+            .map();
+
+        Map<String, Object> enrichValueFieldMapping = MapBuilder.<String, Object>newMapBuilder()
+            .put("type", "{dynamic_type}")
+            .put("indexed", false)
+            .put("doc_values", true)
+            .map();
+
+        Map<String, Object> enrichValueTemplateDefinition = MapBuilder.<String, Object>newMapBuilder()
+            .put("match", "*")
+            .put("unmatch", policy.getEnrichKey())
+            .put("mapping", enrichValueFieldMapping)
+            .map();
+
+        Map<String, Object> enrichValueTemplate = MapBuilder.<String, Object>newMapBuilder()
+            .put("enrich_value_template", enrichValueTemplateDefinition)
+            .map();
+
+        List<Map<String, Object>> templates = new ArrayList<>();
```

Comment:
I should have mentioned this yesterday. I'm currently developing the enrich processor and we no longer need define a mapping for each enrich value field separately. I'm working on a custom meta field type, that does something similar that the `SourceFieldMapper` is doing, but instead stores the source of a document in binary doc values. 

The advantage is that at enrich time we only need to do a single doc values lookup instead of multiple. This going to be more efficient in case that there are multiple enrich value fields, which I think will be the case.

The other consequence is that enrich value fields shouldn't be defined in the mapping of an enrich index. The mapping will look something like:

```
{
   "_source": {"enabled": false}
   "_enrich_source": {"enabled: true}
   "properties": {
       "enrich_key": {
          "type": "keyword",
          "doc_values": false
       }
   }
}
```

Reply:
This makes things so much more simple, thanks.